Source/src/components/ComponentBillboard.cpp

```diff
@@ -0,0 +1,549 @@
+{
+    const std::string title = "Select billboard texture ";
+    std::string texture_path;
+    ResourceTexture* res = nullptr;
+
+    if (ImGui::Button("Add Texture"))
+    {
+        ImGuiFileDialog::Instance()->OpenDialog(title.c_str(),
+            "Select Texture", ".png,.tif,.jpg,.tga", "./assets/textures/", 1, nullptr,
+            ImGuiFileDialogFlags_DontShowHiddenFiles | ImGuiFileDialogFlags_DisableCreateDirectoryButton 
+            | ImGuiFileDialogFlags_HideColumnType | ImGuiFileDialogFlags_HideColumnDate);
+    }
+
+    if (ImGuiFileDialog::Instance()->Display(title.c_str()))
+    {
+        if (ImGuiFileDialog::Instance()->IsOk())
+        {
+            texture_path = ImGuiFileDialog::Instance()->GetFilePathName();
+        }
+
+        ImGuiFileDialog::Instance()->Close();
+    }
+
+    texture_path.append(META_EXTENSION);
+    if (!FileSystem::Exists(texture_path.c_str()))
+    {
+        return;
+    }
+
+    YAML::Node texture_node = YAML::LoadFile(texture_path);
+    res = static_cast<ResourceTexture*>(App->resources->GetResource
+    (Resource::Type::TEXTURE, texture_node[RESOURCES][0][RESOURCE_ID].as<UID>()));
+
+    if (res != nullptr)
+    {
+        texture = res;
+        textureID = texture->GetID();
+        has_texture = 1;
+    }
+}
+
+void Hachiko::ComponentBillboard::RemoveTexture()
+{
+    texture = nullptr;
+    textureID = 0;
+}
+
+inline void Hachiko::ComponentBillboard::UpdateAnimationData() 
+{
+    if (!has_texture)
+    {
+        return;
+    }
+
+    if (animation_index.x < x_tiles - 1)
+    {
+        animation_index.x += 1.0f;
+        return;
+    }
+    else if (animation_index.y < y_tiles - 1)
+    {
+        animation_index.x = 0.0f;
+        animation_index.y += 1.0f;
+        return;
+    }
+
+    animation_index = {0.0f, 0.0f};
+}
+
+inline void Hachiko::ComponentBillboard::UpdateColorOverLifetime()
+{
+    float timeMod = fmod(time, billboard_lifetime / color_cycles);
+    color_frame = timeMod / billboard_lifetime * color_cycles;
+
+    if (color_loop && time > billboard_lifetime)
+    {
+        time = 0.0f;
+        color_frame = 0.0f;
+    }
+}
+
+void Hachiko::ComponentBillboard::PublishIntoScene()
+{
+    auto scene = game_object->scene_owner;
+
+    if (scene)
+    {
+        scene->AddParticleComponent(this);
+        in_scene = true;
+    }
+}
+
+void Hachiko::ComponentBillboard::DetachFromScene()
+{
+    auto scene = game_object->scene_owner;
+
+    if (scene)
+    {
+        scene->RemoveParticleComponent(GetID());
+        in_scene = false;
+    }
+}
+
+void Hachiko::ComponentBillboard::GetOrientationMatrix(ComponentCamera* camera, float4x4& model_matrix)
+{
+    ComponentTransform* transform = GetGameObject()->GetComponent<ComponentTransform>();
+
+    Frustum* frustum = camera->GetFrustum();
+    float4x4* proj = &camera->GetProjectionMatrix();
+    float4x4* view = &camera->GetViewMatrix();
+
+    float3x3 rotatePart = transform->GetGlobalMatrix().RotatePart();
```

Comment:
Let's use snake case for naming; `rotate_part`.

Reply:
Done